MM-15128 - add support for interactive dialogs with no elements

andresoro opened the pull request:
<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository specific documentation at https://developers.mattermost.com/contribute
-->

#### Summary
I think this simple change is what is needed to support interactive dialogs with no elements. Since my react/js knowledge is a little limited, I'm not sure if we need to check if the elements array is non null before we render or not. 

#### Ticket Link
Part of [MM-15128](https://github.com/mattermost/mattermost-server/issues/10686)
#### Related Pull Requests

- Has server changes [Pull request #10755](https://github.com/mattermost/mattermost-server/pull/10755)

hanzei commented:
Hi @andresoro,

Thanks for the PR. :+1: 

I've testing it with the demo plugin. See https://github.com/mattermost/mattermost-plugin-demo/pull/28 for the PR that I used.

After creating the dialog with `/dialog no-elements` the web app crashes and I see the following error in the dev console:
```
Unhandled promise rejection TypeError: "props.elements is null"
    InteractiveDialog interactive_dialog.jsx:36
```

Let me know, if you need further information on how to debug this or have other questions.

andresoro commented:
Hey @hanzei it looks like I need to add a check to see if the elements array is null or not. I suspected something like this would happen. I appreciate the update for the demo plugin and I am going to use that to test these changes further. Going to continue to work on this.

andresoro commented:
I used the plugin mentioned above to test and debug the component. Checking for null in the constructor and the render is what's needed to display the interactive dialog properly. The "Confirm" button wasn't doing anything but I'm not sure if that's by design with the plugin or with the dialog component. Also, it seems like there is a blank space left over from Modal.Body since there are no elements. 

It would look cleaner if the Modal.Body was dropped when the elements array is null.

hanzei commented:
Thanks for the update! The dialog opens now!

> The "Confirm" button wasn't doing anything but I'm not sure if that's by design with the plugin or with the dialog component. 

The problem is caused by a TypeError:
```
Unhandled promise rejection TypeError: "elements is null"
    _callee$ interactive_dialog.jsx:56
```

Would you please fix this? Then we can queue this PR for PM review.

andresoro commented:
Went ahead and fixed that error and also changed the Modal.Body to display only if there are elements. I'm not sure if the way I did it was proper or hackish since my react/JS knowledge is rusty.

hanzei commented:
@andresoro Thanks for the update. This works great! One visual problem I've noticed is that `modal-header` has a `border-bottom` and `.modal-footer` has a `border-top`. This makes up for a doubled border. Would you be open on changing this that it only shows one border?

andresoro commented:
Two borders would be the expected behavior when we have elements to display, correct? So we should only have one border when there are no elements.

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

hanzei commented:
@andresoro I see your points. Let's ask our Product Manager @jasonblais what he thinks about this.

jasonblais commented:
Thanks for the contribution! Happy to take a look at it and see how it looks with two borders when there are no elements.

Looks like the build failed, can you help take a look at it @andresoro when you have the chance?

andresoro commented:
Hey @jasonblais, I ran the commands that the CI ran and it works on my machine. I also did some searching and found that someone else had this same issue on [mattermost-mobile #2405](https://github.com/mattermost/mattermost-mobile/issues/2405). Unsure how to proceed with this one.

cpanato commented:
some lint errors
```
/root/mattermost-webapp/components/interactive_dialog/interactive_dialog.jsx
   68:16  error  Trailing spaces not allowed  no-trailing-spaces
   70:1   error  Trailing spaces not allowed  no-trailing-spaces
  192:1   error  Trailing spaces not allowed  no-trailing-spaces
```

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: https://i-0aa9d0e8366ab00ca.test.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Internal IP: 172.26.23.149
Instance ID: i-0aa9d0e8366ab00ca

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

jasonblais commented:
@cpanato The generated test server wasn't loading for me when I tried to access it. I've tried to generate a new one, but it seems to fail.

Next week, please help take a look?

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: https://i-07fd5f252b2116062.test.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Internal IP: 172.26.15.204
Instance ID: i-07fd5f252b2116062

jasonblais commented:
@andreistanciu24 @hanzei Following the previous conversation about having a "double" border on a dialog with no elements:

Is there a way to check that if there are no elements, then do not render the `border-bottom` for `modal-header`, for instance?

![image](https://user-images.githubusercontent.com/13119842/57225923-b2cfe200-6fdb-11e9-90bb-0f7cb8a44d6d.png)

Been testing, and otherwise it's looking really good 👍

jasonblais commented:
Apologies, looks like I at-mentioned the wrong username..re-posting below:

@andresoro @hanzei Following the previous conversation about having a "double" border on a dialog with no elements:

Is there a way to check that if there are no elements, then do not render the `border-bottom` for `modal-header`, for instance?

![image](https://user-images.githubusercontent.com/13119842/57225923-b2cfe200-6fdb-11e9-90bb-0f7cb8a44d6d.png)

mattermod commented:
Spinmint test running for more than 7 days. This test server was terminated.

andresoro commented:
So I did a bit of research on this and found that React allows for [inline styling](https://reactjs.org/docs/dom-elements.html#style) within components. The docs say it is not the _best_ way to style components but it also allows for conditional styling based on the components props. In our case it's whether or not there are elements in the dialog. 

So I modified the Modal.Header tag to the following: 
```
<Modal.Header closeButton={true} style={elements == null && {borderBottom: '0px'}}>
```

Which essentially changes the border-bottom attribute to not show if there aren't elements. I tested with `/dialog no-elements` and the output looks like: 

<img width="669" alt="Screen Shot 2019-05-18 at 8 12 59 PM" src="https://user-images.githubusercontent.com/11357655/57976329-3958cc00-79ab-11e9-987a-deb5668be52f.png">

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: https://i-02f16f0469e1316fa.test.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Internal IP: 172.26.26.24
Instance ID: i-02f16f0469e1316fa

mattermod commented:
Spinmint test server destroyed

hanzei commented:
@asaadmahmood @cpoile Gentle reminder to review this PR

cpoile commented on components/interactive_dialog/interactive_dialog.jsx:
```diff
@@ -51,18 +53,20 @@ export default class InteractiveDialog extends React.Component {
-                );
-            }
-        });
+        if (elements != null) {
```
Suggest: `if (elements) {` 

which would help protect against problems with undefined, null, and type coercion

cpoile commented on components/interactive_dialog/interactive_dialog.jsx:
```diff
@@ -160,15 +164,18 @@ export default class InteractiveDialog extends React.Component {
                     </Modal.Title>
                 </Modal.Header>
-                <Modal.Body>
+                {elements && <Modal.Body>
```
👍

andresoro commented:
@cpoile thanks for the review. I'll make the changes and push 👍.

cpoile commented:
Great, just remember to click request re-review so I know to come back. Thanks!
